src/climate_learn/loaders.py

```diff
@@ -0,0 +1,356 @@
+):
+    if len(list(net.parameters())) == 0:
+        warnings.warn(
+            "Net has no trainable parameters, setting optimizer to `None`"
+        )
+        optimizer = None
+    if optim.lower() == "sgd":
+        optimizer = torch.optim.SGD(net.parameters(), **optim_kwargs)
+    elif optim.lower() == "adam":
+        optimizer = torch.optim.Adam(net.parameters(), **optim_kwargs)
+    elif optim.lower() == "adamw":
+        optimizer = torch.optim.AdamW(net.parameters(), **optim_kwargs)
+    else:
+        raise NotImplementedError(
+            f"{optim} is not an implemented optimizer. If you think it should"
+            " be, please raise an issue at"
+            " https://github.com/aditya-grover/climate-learn/issues"
+        )
+    return optimizer
+
+def load_lr_scheduler(
+    sched: str,
+    optimizer: torch.optim.Optimizer,
+    sched_kwargs: Dict[str, Any] = {}
+):
+    if optimizer is None:
+        warnings.warn(
+            "Optimizer is `None`, setting LR scheduler to `None` too"
+        )
+        lr_scheduler = None
+    if sched == "constant":
+        lr_scheduler = LRScheduler.ConstantLR(optimizer, **sched_kwargs)
+    elif sched == "linear":
+        lr_scheduler = LRScheduler.LinearLR(optimizer, **sched_kwargs)
+    elif sched == "exponential":
+        lr_scheduler = LRScheduler.ExponentialLR(optimizer, **sched_kwargs)
+    elif sched == "linear-warmup-cosine-annealing":
+        lr_scheduler = LinearWarmupCosineAnnealingLR(optimizer, **sched_kwargs)    
+    else:
+        raise NotImplementedError(
+            f"{sched} is not an implemented learning rate scheduler. If you"
+            " think it should be, please raise an issue at"
+            " https://github.com/aditya-grover/climate-learn/issues"
+        )
+    return lr_scheduler
+
+def load_loss(loss_name, metainfo):
+    loss_cls = METRICS_REGISTRY.get(loss_name, None)
+    if loss_cls is None:
+        raise NotImplementedError(
+            f"{loss_name} is not an implemented loss. If you think it should be,"
+            " please raise an issue at"
+            " https://gtihub.com/aditya-grover/climate-learn/issues"
+        )
+    loss = loss_cls(
+        aggregate_only=True,
+        metainfo=metainfo
+    )
+    return loss
+
+def load_transform(transform_name, data_module):
+    transform_cls = TRANSFORMS_REGISTRY.get(transform_name, None)
+    if transform_cls is None:
+        raise NotImplementedError(
+            f"{transform_name} is not an implemented transform. If you think"
+            " it should be, please raise an issue at"
+            " https://github.com/aditya-grover/climate-learn/issues"
+        )
+    transform = transform_cls(data_module)
+    return transform
+        
+def get_data_dims(data_module):
+    for batch in data_module.train_dataloader():
+        x, y, _, _ = batch
+        break
+    return x.shape, y.shape
+
+def get_data_variables(data_module):
+    for batch in data_module.train_dataloader():
```

Comment:
Using `Task` was my initial approach. However, the `in_vars` are dynamic. For example, if the user requests `geopotential` in the task arguments, then this code would think there's just one input variable. In reality, the geopotentials for every pressure level are loaded. I believed that changing the task argument code would be beyond the scope of this PR, since it's a problem from the data side.

Reply:
Actually when the `.setup()`[[See this](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/data/dataset/map_dataset.py#L105)] is called for the `MapDataset` and `ShardDataset`, it first calls the `.setup()` [[See this](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/data/dataset/map_dataset.py#L106)] for `climate_dataset` which returns `variables_to_update`.

`variables_to_update` is a dict where the keys are the old variable names (like `geopotential` in this case) and values are a list of new variable names corresponding to the old one (like `['geopotential_50', 'geopotential_250', ..., etc]` in this case). Then, the `.setup()` corresponding to the `MapDataset` and `ShardDataset` calls the `.setup()`[[See this](https://github.com/aditya-grover/climate-learn/blob/d6cf91d0bb1dff03e8a19dbfed12764e66fd40da/src/climate_learn/data/dataset/map_dataset.py#L122)] corresponding to the `task`. Which updates the `in_vars`, `out_vars` and `constants` for the `Task`.

Thus, if the `.setup() `has been called for `dataset` then you can safely use the `in_vars`. In fact the current logic would also fail if the `.setup()` hasn't been called because there would be no data in the memory, thus `dataloader` would give an error.